txnkv/transaction/txn.go

```diff
@@ -704,19 +910,68 @@ func (txn *KVTxn) LockKeys(ctx context.Context, lockCtx *tikv.LockCtx, keysInput
 			}
 		}
 		if txn.committer.primaryKey == nil {
-			txn.committer.primaryKey = keys[0]
 			assignedPrimaryKey = true
+			txn.selectPrimaryForPessimisticLock(keys)
+		}
+
+		txn.committer.forUpdateTS = lockCtx.ForUpdateTS
+
+		allKeys := keys
+		// If aggressive locking is enabled and we don't need to update the primary for all locks, we can avoid sending
+		// RPC to those already locked keys.
+		if txn.aggressiveLockingContext != nil {
```

Comment:
How about wrapping this part into a function like `filterAggresiveLockedKeys` since the `LockKeys` is already a little bit long?